lib/semantics/resolve-names.cc

```diff
@@ -4071,16 +4086,8 @@ bool ConstructVisitor::Pre(const parser::LocalitySpec::LocalInit &x) {
 
 bool ConstructVisitor::Pre(const parser::LocalitySpec::Shared &x) {
   for (const auto &name : x.v) {
-    Symbol *prev{FindSymbol(name)};
-    if (!prev) {
-      Say(name, "Variable '%s' not found"_err_en_US);
-      context().SetError(
-          MakeSymbol(name, ObjectEntityDetails{EntityDetails{}}));
-    } else if (prev->owner() == currScope()) {
-      SayAlreadyDeclared(name, *prev);  // C1125 and C1126
-    } else if (!IsVariableName(*prev)) {
-      SayLocalMustBeVariable(name, *prev);  // C1124
-    } else {
+    Symbol *prev{EnsureEnclosingEntity(name)};
+    if (PassesSharedLocalityChecks(name, *prev)) {
       auto &symbol{MakeSymbol(name, HostAssocDetails{*prev})};
       symbol.set(Symbol::Flag::LocalityShared);
       name.symbol = &symbol;  // override resolution to parent
```

Comment:
I know this is not completely part of your change, but why is:

```
if (auto *type{prev->GetType()}) {
    symbol.SetType(*type);
    if (prev->test(Symbol::Flag::Implicit)) {
      symbol.set(Symbol::Flag::Implicit);
    }
  }
```
performed in `DeclareLocalEntity` for LOCAL and LOCAL_INIT variables and not here for SHARED variables ?

Reply:
Good question.  After consultation with @tskeith, I don't think that this code is needed in either case.  I'll remove it.

Reply:
The `ApplyImplicitRules` part was not needed in `DeclareLocalEntity` but copying the type and implicit flag still are.

The same code isn't needed here because the new symbol is a host-association with the previous one. The type of `symbol` should come from the type of `prev`.